File: solutions/csharp/nucleotide-count/1/NucleotideCount.cs
using System;
using System.Collections.Generic;

public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    {
        IDictionary<char, int> nucleotideMap = new Dictionary<char, int>
        {
            ['A'] = 0,
            ['C'] = 0,
            ['G'] = 0,
            ['T'] = 0
        };

        foreach (char nucleotide in sequence)
        {
            if (!nucleotideMap.ContainsKey(nucleotide))
            {
                throw new ArgumentException();
            }

            nucleotideMap[nucleotide]++;
        }

        return nucleotideMap;
    }
}

File: solutions/cpp/nucleotide-count/1/nucleotide_count.cpp
#include "nucleotide_count.h"
#include <stdexcept>

namespace nucleotide_count {

std::map<char, int> count(const std::string& dna_chain) {
    std::map<char, int> nucleotide_map{{'A', 0}, {'C', 0}, {'G', 0}, {'T', 0}};

    for (char nucleotide : dna_chain) {
        switch (nucleotide) {
            case 'A': case 'C': case 'G': case 'T':
                nucleotide_map[nucleotide]++;
                break;
            default:
                throw std::invalid_argument("Invalid DNA string.");
        }
    }

    return nucleotide_map;
}

} // namespace nucleotide_count